[5.1-04-24-2019][Reflection] Switch Remote Mirror back to uintptr_t for pointers on watchOS.

mikeash opened the pull request:
Cherry-pick https://github.com/apple/swift/pull/24529 for 5.1-04-024-2019.

Existing Remote Mirror dylibs are compiled this way, and changing it breaks binary compatibility. We want to use uint64_t everywhere else, since the target's pointer size may not match ours.

rdar://problem/50279443

mikeash committed:
[Reflection] Switch Remote Mirror back to uintptr_t for pointers on watchOS.

Existing Remote Mirror dylibs are compiled this way, and changing it breaks binary compatibility. We want to use uint64_t everywhere else, since the target's pointer size may not match ours.

rdar://problem/50279443

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

bob-wilson commented:
I would be happy to approve this but it would be good to get someone who knows the remote mirror code to review it. Maybe @slavapestov or @dcci?

dcci reviewed (approved):
LGTM.

dcci commented:
(although it's a little sad we have to play these tricks).

mikeash commented:
@swift-ci please test

mikeash committed:
[Reflection] Include TargetConditionals.h for WatchOS detection, not Availability.h.

While we're in there, make SwiftReflectionTest's debugLog function take an @autoclosure so we don't waste a ton of time constructing log messages that are never logged.

mikeash commented:
@swift-ci please test
